File: client/csharp/example/connect/test/cli/DeviceMenu.cs
using System;
using Connect;

namespace example
{
  class DeviceMenu {
    private Menu menu;
    private static ConnectSvc connectSvc;

    public DeviceMenu(ConnectSvc svc) {
      connectSvc = svc;

      MenuItem[] items = new MenuItem[7];
      items[0] = new MenuItem("1", "Set connection mode", new MenuCallback(SetConnectionMode), false);
      items[1] = new MenuItem("2", "Enable SSL", new MenuCallback(EnableSSL), false);
      items[2] = new MenuItem("3", "Disable SSL", new MenuCallback(DisableSSL), false);
      items[3] = new MenuItem("4", "Disconnect devices", new MenuCallback(Disconnect), false);
      items[4] = new MenuItem("5", "Disconnect all devices", new MenuCallback(DisconnectAll), false);
      items[5] = new MenuItem("6", "Refresh the managed device list", new MenuCallback(ShowDeviceList), false);
      items[6] = new MenuItem("q", "Return to Main Menu", null, true);      

      menu = new Menu(items);
    }

    public void Show() {
      if(GetDeviceList() == 0) {
        Console.WriteLine("No connected device. Please connect to some devices first.");
        return;
      }

      menu.Show("Device Menu");
    }

    public static int GetDeviceList() {
      Console.WriteLine("Getting the devices managed by the gateway...");

      try {
        var devList = connectSvc.GetDeviceList();
        
        Console.WriteLine();
        Console.WriteLine("***** Managed Devices: {0}", devList.Count);
        for(int i = 0; i < devList.Count; i++) {
          Console.WriteLine(devList[i]);
        }

        Console.WriteLine();

        return devList.Count;
      } catch(Exception e) {
        Console.WriteLine("Cannot get device list: {0}", e);
        return 0;
      }
    }

    public static void ShowDeviceList() {
      GetDeviceList();
    }

    public static void SetConnectionMode() {
      Console.WriteLine("Enter the device IDs to to change the connection mode");
      
      uint[] deviceIDs = Menu.GetDeviceIDs();

      if(deviceIDs == null) {
        return;
      }

      Console.Write(">> Select the connection mode (0: Gateway to Device(default), 1: Device to Gateway): ");
      string modeStr = Console.ReadLine();

      int mode;
      if(!Int32.TryParse(modeStr, out mode)) {
        Console.Write("Invalid connection mode: {0}", modeStr);
      }

      if(mode != 0 && mode != 1) {
        Console.Write("Invalid connection mode: {0}", modeStr);
      }

      Console.WriteLine("Changing the connection mode...");

      try {
        connectSvc.SetConnectionMode(deviceIDs, (ConnectionMode)mode);
      } catch(Exception e) {
        Console.WriteLine("Cannot set connection mode: {0}", e);        
      }
    }    



    public static void EnableSSL() {
      Console.WriteLine("Enter the device IDs to enable");

      uint[] deviceIDs = Menu.GetDeviceIDs();

      if(deviceIDs == null) {
        return;
      }

      Console.WriteLine("Enabling SSL...");

      try {
        connectSvc.EnableSSL(deviceIDs);
      } catch(Exception e) {
        Console.WriteLine("Cannot enable SSL: {0}", e);        
      }
    }    

    public static void DisableSSL() {
      Console.WriteLine("Enter the device IDs to disable");

      uint[] deviceIDs = Menu.GetDeviceIDs();

      if(deviceIDs == null) {
        return;
      }

      Console.WriteLine("Disabling SSL...");

      try {
        connectSvc.DisableSSL(deviceIDs);
      } catch(Exception e) {
        Console.WriteLine("Cannot disable SSL: {0}", e);        
      }
    }    

    public static void Disconnect() {
      Console.WriteLine("Enter the device IDs to disconnect");

      uint[] deviceIDs = Menu.GetDeviceIDs();

      if(deviceIDs == null) {
        return;
      }

      Console.WriteLine("Disconnecting devices...");

      try {
        connectSvc.Disconnect(deviceIDs);
      } catch(Exception e) {
        Console.WriteLine("Cannot disconnect devices: {0}", e);        
      }
    }

    public static void DisconnectAll() {
      Console.WriteLine("Disconnecting all devices...");

      try {
        connectSvc.DisconnectAll();
      } catch(Exception e) {
        Console.WriteLine("Cannot disconnect all devices: {0}", e);        
      }
    }
  }
}

File: client/node/example/connect/test/cli/deviceMenu.js
const menu = require('../../../cli/menu');
const connect = require('../../../connect');
const mainMenu = require('./mainMenu');
const readlineSync = require('readline-sync');

async function showDeviceList() {
  try {
    var devList = await connect.getDeviceList();

    if(devList.length == 0) {
      console.error('No connected device. Connect to some devices first.');
      return null;
    }

    console.log(`\n***** Managed devices: ${devList.length}`);
    console.log(devList);

    return devList;
  }  catch (err) {
    console.error('Cannot get the device list: ', err);
    return null;
  }
}

async function setConnectionMode() {
  console.log('\nEnter the device IDs to change the mode');
  
  var deviceIDs = menu.getDeviceIDs();

  if(deviceIDs.length == 0) {
    console.error('No device to change');
    return;
  }

  var modeStr = readlineSync.question('>> Select the connection mode (0: Gateway to Device(default), 1: Device to Gateway): ');
  var mode = Number(modeStr);

  if(isNaN(mode) || mode != 1) {
    mode = 0;
  } 

  console.log(`Changing the connection mode to ${mode} for ${deviceIDs.length} devices...`);

  try {
    await connect.setConnectionMode(deviceIDs, mode);
  } catch (err) {
    console.error('Cannot change the connection mode: ', err);
  }
}


async function enableSSL() {
	console.log('\nEnter the device IDs to enable SSL');

  var deviceIDs = menu.getDeviceIDs();

  if(deviceIDs.length == 0) {
    console.error('No device to enable');
    return;
  }

  console.log('Enabling SSL...');

  try {
    await connect.enableSSL(deviceIDs);
  } catch (err) {
    console.error('Cannot enable SSL: ', err);
  }
}


async function disableSSL() {
	console.log('\nEnter the device IDs to disable SSL');

  var deviceIDs = menu.getDeviceIDs();

  if(deviceIDs.length == 0) {
    console.error('No device to disable');
    return;
  }

  console.log('Disabling SSL...');

  try {
    await connect.disableSSL(deviceIDs);
  } catch (err) {
    console.error('Cannot disable SSL: ', err);
  }
}


async function disconnect() {
	console.log('\nEnter the device IDs to disconnect');

  var deviceIDs = menu.getDeviceIDs();

  if(deviceIDs.length == 0) {
    console.error('No device to disconnect');
    return;
  }

  console.log('Disconnecting devices...');

  try {
    await connect.disconnect(deviceIDs);
  } catch (err) {
    console.error('Cannot disconnect: ', err);
  }
}

async function disconnectAll() {
  console.log('Disconnecting all devices...');

  try {
    await connect.disconnectAll();
  } catch (err) {
    console.error('Cannot disconnect all: ', err);
  }
}

async function showMenu() {
  var devList = await showDeviceList();

  if(!devList) {
    mainMenu.showMenu();
    return;
  }

  var menuItems = [];
  
  menuItems.push(new menu.MenuItem('1', 'Set connection mode', setConnectionMode, false));
  menuItems.push(new menu.MenuItem('2', 'Enable SSL', enableSSL, false));
  menuItems.push(new menu.MenuItem('3', 'Disable SSL', disableSSL, false));
  menuItems.push(new menu.MenuItem('4', 'Disconnect', disconnect, false));
  menuItems.push(new menu.MenuItem('5', 'Disconnect All', disconnectAll, false));
  menuItems.push(new menu.MenuItem('6', 'Refresh the device list', showDeviceList, false));
  menuItems.push(new menu.MenuItem('q', 'Return to Main Menu', mainMenu.showMenu, true));

  menu.showMenu('Device Menu', menuItems)  
}

module.exports.showMenu = showMenu;
